src/providers/autocomplete/AutocompleteSources.hpp

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include "common/Channel.hpp"
+#include "messages/Emote.hpp"
+#include "providers/autocomplete/AutocompleteSource.hpp"
+
+#include <QString>
+
+#include <functional>
+#include <memory>
+#include <utility>
+#include <variant>
+
+namespace chatterino {
+
+//// AutocompleteEmoteSource
+
+struct CompletionEmote {
+    /// Emote image to show in input popup
+    EmotePtr emote{};
+    /// Name to check completion queries against
+    QString searchName{};
+    /// Name to insert into split input upon tab completing
+    QString tabCompletionName{};
+    /// Display name within input popup
+    QString displayName{};
+    /// Emote provider name for input popup
+    QString providerName{};
+    /// Whether emote is emoji
+    bool isEmoji{};
+};
+
+class AutocompleteEmoteSource : public AutocompleteSource
+{
+public:
+    using ActionCallback = std::function<void(const QString &)>;
+    using AutocompleteEmoteStrategy = AutocompleteStrategy<CompletionEmote>;
+
+    /// @brief Initializes a source for CompletionEmotes from the given channel
+    /// @param channel Channel to initialize emotes from
+    /// @param strategy AutocompleteStrategy to apply
+    /// @param callback ActionCallback to invoke upon InputCompletionItem selection.
+    /// See InputCompletionItem::action(). Can be nullptr.
+    AutocompleteEmoteSource(const Channel &channel,
```

Comment:
This should just take a `const Channel *channel` - it currently goes from a shared_ptr<Channel>, gets deferenced to this reference, then internally gets referenced to a pointer